### PR TITLE
BUG: setting dt64 values into Series[int] incorrectly casting dt64->int

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -991,6 +991,18 @@ def setitem(self, indexer, value):
...
+                self.is_object
+                and not is_ea_value
+                and arr_value.dtype.kind in ["m", "M"]
+            ):
+                # https://github.com/numpy/numpy/issues/12550
+                #  numpy will incorrect cast to int if we're not careful
+                if is_list_like(value):
+                    value = list(value)
+                else:
+                    value = [value] * len(values[indexer])
```
Comment (jreback) [inline pandas/core/internals/blocks.py:1004]: maybe use `np.full`?

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -991,6 +991,18 @@ def setitem(self, indexer, value):
...
         else:
+            if (
+                self.is_object
+                and not is_ea_value
+                and arr_value.dtype.kind in ["m", "M"]
+            ):
+                # https://github.com/numpy/numpy/issues/12550
+                #  numpy will incorrect cast to int if we're not careful
+                if is_list_like(value):
+                    value = list(value)
```
Comment (jreback) [inline pandas/core/internals/blocks.py:1002]: `np.asarray`?

File: pandas/core/internals/blocks.py [hunk 3]
```diff
@@ -991,6 +991,18 @@ def setitem(self, indexer, value):
 
         # set
         else:
+            if (
```
Comment (jreback) [inline pandas/core/internals/blocks.py:994]: elif?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:1002]: until numpy#12550 is fixed this is the only way i found to get this to work

File: pandas/core/internals/blocks.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:994]: no, we still want to fall through to L1006

File: pandas/core/internals/blocks.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:1004]: same as above re numpy#12550

File: pandas/core/internals/blocks.py (hunk 3 above)
Comment (jreback) [reply pandas/core/internals/blocks.py:994]: sure but i thin it is much more clear to use an elif and then simply repeat that line in the else

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
> sure but i thin it is much more clear to use an elif and then simply repeat that line in the else

ok, will update.

BTW #38709, #39044, #39163 also touch this code, will need to get those merged/rejected in order to move forward on the refactor in #39302.

(and a very gentle ping on #38992 since I want to write more tests)

File: pandas/core/internals/blocks.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py]: updated + green

**Comment by jbrockmendel** [comment]
rebased + green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. can you merge master and ping on green

**Pull request merged**